### ISSUE TITLE
如果分割的目标区域在整张图上占比特别小，导致样本严重不均衡，如何选择损失函数

### ISSUE THREAD

**Issue by wenzhouwww** [zhixuhao/unet#101]
在用到这一模型做语义分割的时候，由于要分割的语义区域在图片中特别小，从而使得非语义区域，也会就是mask=0的像素点占的比例过大，从而使得二值交叉熵  binary_crossentropy作为损失函数，会使得模型偏向于负样本，模型效果非常差，不知道有没改善的方式？

**Comment by huangy94** [comment]
或许可以试试dice_loss？

**Comment by wenzhouwww** [comment]
从某种程度上讲，dice loss是一种有效的策略，但面对样本严重不均衡，正负样本比例1:1000左右的这种情况，dice loss感觉应该不是很好用，毕竟dice loss的策略讲的是让正负样本的轮廓与ground truth 接近，体现的是轮廓相似程度，我个人这么理解的，网络目前还没有用dice loss跑过，不知道您有做过dice loss作为损失函数，来训练语义分割的工作么，可否给点建议。

**Comment by jizhang02** [comment]
我使用了dice loss，但是分割出的结果是全白的？这是为什么？

**Comment by wenzhouwww** [comment]
建议你看看网络预测出来的输出，也就是预测出来的那个numpy数组里的数值。另外，研究网络的loss函数是不是有问题

**Comment by jizhang02** [comment]
> 建议你看看网络预测出来的输出，也就是预测出来的那个numpy数组里的数值。另外，研究网络的loss函数是不是有问题

 def dice_coef(y_true, y_pred):
     smooth = 1
     y_true_f = K.flatten(y_true)
     y_pred_f = K.flatten(y_pred)
     intersection = K.sum(y_true_f * y_pred_f)
     return (2. * intersection +smooth) / (K.sum(y_true_f) + K.sum(y_pred_f) +smooth)

 def dice_coef_loss(y_true, y_pred):
     print("dice loss")
     return 1-dice_coef(y_true, y_pred)
...
model.compile(optimizer = Adam(lr = 1e-4), loss = dice_coef_loss, metrics = ['accuracy'])
我这样使用了dice损失函数，您看有什么问题吗？

**Comment by WangHaiXu1** [comment]
> 建议你看看网络预测出来的输出，也就是预测出来的那个numpy数组里的数值。另外，研究网络的loss函数是不是有问题

我同样有这个问题，之前一直输出全黑，年前调了一些东西后好了，也不清楚改了什么，现在又是这个样子

**Comment by wenzhouwww** [comment]
你仔细看下dice_coef，你会发现，如果你的分割目标倘若是1，你假设为1的像素点数为10个，而一个batchsize中假设输入一张图，整张图中像素点个数为500个， 你自己计算一下使用dice系数作为loss函数的话，如果这10个像素点都是预测成了0，它的损失是多大，这样你可以推测面对样本不均衡，你使用dice losss是否合理。

**Comment by wenzhouwww** [comment]
> > 建议你看看网络预测出来的输出，也就是预测出来的那个numpy数组里的数值。另外，研究网络的loss函数是不是有问题
> 
> 我同样有这个问题，之前一直输出全黑，年前调了一些东西后好了，也不清楚改了什么，现在又是这个样子

输出全黑意味着全都预测成了0，可能是模型跑偏了，检查样本是否均衡，另外，损失函数的选取也需要考虑下。

**Comment by wenzhouwww** [comment]
> 你仔细看下dice_coef，你会发现，如果你的分割目标倘若是1，你假设为1的像素点数为10个，而一个batchsize中假设输入一张图，整张图中像素点个数为500个， 你自己计算一下使用dice系数作为loss函数的话，如果这10个像素点都是预测成了0，它的损失是多大，这样你可以推测面对样本不均衡，你使用dice losss是否合理。
从某种程度讲，dice_loss更偏向于分割目标，也就是我们的这10个像素点

**Comment by jizhang02** [comment]
> 你仔细看下dice_coef，你会发现，如果你的分割目标倘若是1，你假设为1的像素点数为10个，而一个batchsize中假设输入一张图，整张图中像素点个数为500个， 你自己计算一下使用dice系数作为loss函数的话，如果这10个像素点都是预测成了0，它的损失是多大，这样你可以推测面对样本不均衡，你使用dice losss是否合理。

谢谢您，听君一席话，开朗不少。望今后再交流！

**Comment by Major-Yuan** [comment]
you can try weighted_binary_crossentropy.

**Comment by ihunhh** [comment]
focal loss

**Comment by FengZhiheng** [comment]
> > 建议你看看网络预测出来的输出，也就是预测出来的那个numpy数组里的数值。另外，研究网络的loss函数是不是有问题
> 
> 我同样有这个问题，之前一直输出全黑，年前调了一些东西后好了，也不清楚改了什么，现在又是这个样子

我有相同的问题，网络的输出全都是黑的，做的是图像分割任务。现在一头雾水。sad

**Comment by Easonrust** [comment]
I have the same problem, and when I change the loss function to 'dice coef loss', my problem was fixed

**Comment by XUYUNYUN666** [comment]
> I have the same problem, and when I change the loss function to 'dice coef loss', my problem was fixed

Can you show me the code of the dice loss function? thank you

**Comment by XUYUNYUN666** [comment]
> 或许可以试试dice_loss？

请问你实现了dice  loss吗？有代码分享一下吗

**Comment by congxxx** [comment]
> > 你仔细看下dice_coef，你会发现，如果你的分割目标倘若是1，你假设为1的像素点数为10个，而一个batchsize中假设输入一张图，整张图中像素点个数为500个， 你自己计算一下使用dice系数作为loss函数的话，如果这10个像素点都是预测成了0，它的损失是多大，这样你可以推测面对样本不均衡，你使用dice losss是否合理。
> > 从某种程度讲，dice_loss更偏向于分割目标，也就是我们的这10个像素点

哥，你解决了吗？我也遇到这个问题了
